core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -231,6 +216,24 @@ private Object handleTxCommand(TxInvocationContext ctx, TransactionBoundaryComma
          }
       } else {
          if (currentTopologyId() > command.getTopologyId()) {
+            /*
+             * We need to retry in case of topology-id mismatch.
+             * We rollback the transaction is it is a 2PC. The rollback is safe (even with state transfer prepared transactions)
+            because we ask the coordinator
+            to retry the transaction. The retry will re-acquire the locks and perform the validation needed. If this node
```

Comment:
To clarify, there is no test for this, I only (ab)used Matej's stress tests to confirm that the problem didn't reappear with my commit. However, my fix relies on `PrepareCommand` being idempotent, and I believe your change would break that guarantee.